build-system/babel-plugins/babel-plugin-transform-annotation-to-extern/index.js

```diff
@@ -0,0 +1,102 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Global typedef map typedefName: typedef comment
+const TYPEDEFS = new Map();
+
+const buildTypedefs = (t, path) => {
+  const typedefs = [];
+  for (const [typedefName, typedefComment] of TYPEDEFS) {
+    const ast = buildVarDeclAndComment(t, path, typedefName, typedefComment);
+    typedefs.push(ast);
+  }
+  return typedefs;
+};
+
+const buildVarDeclAndComment = (t, path, name, comment) => {
+  const decl = t.variableDeclaration('let', [
+    t.variableDeclarator(t.identifier(name)),
+  ]);
+  // Add the typedef annotation.
+  t.addComment(decl, 'leading', comment);
+  return decl;
+};
+
+module.exports = function(babel) {
+  const {types: t} = babel;
+  // `shouldWriteToFile` should only be true in the production pipeline.
+  let shouldWriteToFile;
+  // `shouldEmitTypedefs` is used for testing purposes only.
+  let shouldEmitTypedefs;
+  return {
+    pre() {
+      TYPEDEFS.clear();
+      const {writeToFile = false, emitTypedefs = false} = this.opts;
+      shouldWriteToFile = writeToFile;
+      shouldEmitTypedefs = emitTypedefs;
+    },
+    visitor: {
+      Program: {
+        exit(path) {
+          if (shouldWriteToFile) {
+            // Stub
+            console.log('write to temp directory');
+          }
+          if (shouldEmitTypedefs) {
+            // Preserve the leading LICENSE comment.
+            path.addComment('leading', path.parent.comments[0].value);
```

Comment:
This is only for tests, why add a trailing comment and leave the program intact?

Reply:
do you mean to not replace the program or to just not add the comment? if the latter its mostly to appease the linter

Reply:
Don't do anything to the program, just postpend the typedefs:

```js
path.addComment('trailing', typdefsComment)
```

Reply:
done.